### DISCUSSION TITLE
Still have question with MiXCR gene library

### DISCUSSION THREAD

**Discussion by omegahh** [milaboratory/mixcr#1761] [Q&A] [answered]
> MiXCR v4.7.0 (built Thu Aug 08 03:19:48 CST 2024; rev=976ba14139; branch=no_branch; host=fv-az1019-185)
> RepSeq.IO v2.5.0 (rev=06fa1852ee)
> MiLib v3.5.0 (rev=b6cfcdc2af)
> Built-in V/D/J/C library: repseqio.v5.1
> 
> Library search path:
> - built-in libraries
> - /home/hongh/.

I export MiXCR library by the `mixcr buildLibrary --species hs ....` command. But when I parsing the output json file, I am confused with the 'alleleInfo' data.

Let me explain with some examples:

### Example 1

As shown in below, the mixcr library tells me that "IGHV1-2\*02" is an allele of "IGHV1-2\*04" with mutation "ST198A". **Am I right?**
<img width="563" alt="截屏2024-08-23 15 47 13" src="https://github.com/user-attachments/assets/949681bf-767c-474c-bec6-a65cd9db1ac3">

Then I check them in the **IMGT Library**, They are "CAGGGC**T**" and "CAGGGC**A**" as shown in below. This is consistent with the information above.

<img width="1460" alt="截屏2024-08-23 16 00 18" src="https://github.com/user-attachments/assets/fa86ceb8-f324-465a-bafd-37ac3fdc4825">

But when I go back to the _Gene Library_ (https://vdj.online/library), The sequence remains "CAGGGC**T**", which **is quite perplexing to me.**

<img width="1588" alt="截屏2024-08-23 15 59 09" src="https://github.com/user-attachments/assets/c14a01de-6860-4c5a-8196-143229094f18">

---

### Example 2

When I test mutations with insertions and deletions, I found something more confusing.

Let's look at gene "IGHV3-30-5\*06x“ and "IGHV3-30\*18", which the latter is the allele parent of the former. And the mutation is "I132GDG135", which means an insertion of 'G' in position 132 and a deletion of 'G' in position 135. Shown in the following:

<img width="565" alt="截屏2024-08-23 20 51 04" src="https://github.com/user-attachments/assets/f138f2d0-305f-4dcf-a3ea-9f0a87bc07c2">

Based on the characteristics of the mutation mentioned above, I believe it should be "CGTG." However, the Gene Library still displays "CTGG", as shown in the following:

<img width="1582" alt="截屏2024-08-23 20 57 14" src="https://github.com/user-attachments/assets/cc7f69eb-432a-42d4-8ee3-b46351bae90a">

Furthermore, I download the fasta file for gene "IGHV3-30-5*06x“ and "IGHV3-30*18", and compare these two sequences. The surprising thing is the different bases are totally different from those reported in the MiXCR library.  

<img width="698" alt="截屏2024-08-23 21 05 40" src="https://github.com/user-attachments/assets/b2ebb64e-ab86-487a-8537-459535c3658d">
<img width="450" alt="截屏2024-08-23 21 08 49" src="https://github.com/user-attachments/assets/31ed95a1-8c20-4215-bf7d-9ef2d7a8593a">

---

### In conclusion

Because I need the detailed sequence information of the aligned genes for downstream analysis, I must know what kind of MiXCR sequence library I am comparing against. This includes all the alleles, as I used the `findAlleles` command, and the exported library is also in this format. Facing these inconsistence, What should I do? How to deal with the exported libraries? Is there any suggestions if I don't want to use the IMGT library?

**Comment by mizraelson** [comment, marked as answer]
Hi, I apologize for the inconvenience. Please refer to the current build in library, as the VDJ online database is outdated. We will update it soon to ensure consistency.

**Comment by omegahh** [comment]
Thank you for replying. So, I would like to confirm the inferred "CGTG" is correct, right? Because for this allele, I can't find it in the IMGT database.

**Comment by mizraelson** [comment]
So, `I132GDG135` means:
- Insertion of a G in the position 132
- Deletion of a G in a position 135.

This leads to:
```
IGHV3-30*18    AAGGGG-CTGGAG
IGHV3-30*5*06  AAGGGGGCTG-AG
```

**Comment by omegahh** [comment]
Thank you, I think I understand what you mean. But I still have a question need to make sure. If there are two continuous insertions, 
<img width="1750" alt="截屏2024-09-06 20 34 18" src="https://github.com/user-attachments/assets/25b41b80-78d9-4ad3-8d15-d01b4c0a8886">

Take gene 'IGHV5-10-1×01' for example, the insertions result should be "CA" or "AC"? which one is the right sequence?

**Reply by mizraelson** [reply]
When several nucleotides are inserted, the order is maintained: `I294CI294A` -> CA.

**Reply by omegahh** [reply]
OK, I see
